cmd/serve.go

```diff
@@ -161,84 +170,83 @@ func doPerf(ctx context.Context, p *node) error {
 	return nil
 }
 
-func listenPerf(ctx context.Context) func(w http.ResponseWriter, r *http.Request) {
-	return func(w http.ResponseWriter, r *http.Request) {
-		// Use this trailer to send additional headers after sending body
-		w.Header().Set("Trailer", "FinalStatus")
-
-		w.Header().Set("Content-Type", "application/octet-stream")
-		w.WriteHeader(http.StatusOK)
-
-		copyContext := func(ctx context.Context, dst io.Writer, src io.Reader) (written int64, err error) {
-			size := 32 * 1024
-			if l, ok := src.(*io.LimitedReader); ok && int64(size) > l.N {
-				if l.N < 1 {
-					size = 1
-				} else {
-					size = int(l.N)
-				}
+func listenPerf(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	// Use this trailer to send additional headers after sending body
+	w.Header().Set("Trailer", "FinalStatus")
+
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.WriteHeader(http.StatusOK)
+
+	copyContext := func(ctx context.Context, dst io.Writer, src io.Reader) (written int64, err error) {
+		size := 32 * 1024
+		if l, ok := src.(*io.LimitedReader); ok && int64(size) > l.N {
+			if l.N < 1 {
+				size = 1
+			} else {
+				size = int(l.N)
 			}
-			buf := make([]byte, size)
-
-		loop:
-			for {
-				select {
-				case <-ctx.Done():
-					return written, ctx.Err()
-				default:
-					// If the reader has a WriteTo method, use it to do the copy.
-					// Avoids an allocation and a copy.
-					if wt, ok := src.(io.WriterTo); ok {
-						return wt.WriteTo(dst)
-					}
+		}
+		buf := make([]byte, size)
 
-					// Similarly, if the writer has a ReadFrom method, use it to do the copy.
-					if rt, ok := dst.(io.ReaderFrom); ok {
-						return rt.ReadFrom(src)
+	loop:
+		for {
+			select {
+			case <-ctx.Done():
+				return written, ctx.Err()
+			default:
+				// If the reader has a WriteTo method, use it to do the copy.
```

Comment:
Can you not use `io.Copy` because this looks like io.Copy https://golang.org/src/io/io.go?s=12796:12856#L384

Reply:
If you need to disconnect the caller based on context you should simply wrap the reader 

```
// Returns error if the cancelCh has been closed (indicating that S3 client has disconnected)
type detectDisconnect struct {
        io.ReadCloser
        cancelCh <-chan struct{}
}

func (d *detectDisconnect) Read(p []byte) (int, error) {
        select {
        case <-d.cancelCh:
                return 0, io.ErrUnexpectedEOF
        default:
                return d.ReadCloser.Read(p)
        }
}
```

So your code will become very simple

```
io.Copy(ioutil.Discard, &detectDisconnect{r.Body, r.Context().Done()})
```

Reply:
So this entire code becomes just few lines.